farm/modeling/prediction_head.py

```diff
@@ -849,66 +881,174 @@ def prepare_labels(self, start_position, end_position, **kwargs):
+            end = r["pred_end"] + r["passage_shift"]
+            return f"{start}-{end}"
+
+        data = {}
+        data["pred_start"] = np.concatenate([x[0] for x in preds])
+        data["pred_end"] = np.concatenate([x[1] for x in preds])
+        data["logit_sum"] = np.concatenate([x[2] for x in preds])
+        data["sample_id"] = np.concatenate([x[3] for x in preds])
+        data["question_shift"] = np.concatenate([x[4] for x in preds])
+        data["passage_shift"] = np.concatenate([x[5] for x in preds])
+        df = pd.DataFrame(data=data)
+        df.loc[:,"answer_indices"] = df.apply(lambda row: create_answeridx_string(row), axis=1)
+
+        # we sometimes have multiple predictions for one sample (= paragraph question pair)
+        # because we split the paragraph into smaller passages
+        # we group all predictions by sample_id
+        unique_sample_ids = df.sample_id.unique()
+        max_per_sample = []
+        for uid in unique_sample_ids:
+            group = df.loc[df.sample_id == uid, :]
+            idx_text_answers = (group.pred_start + group.pred_end) > 0
+            # if we have a text answer in the group we want to discard all no text passages
+            # Reasoning: consider how data is constructed from labels: If we split a document into 5 passages and
+            # only passage no.3 has the text answer, all remaining passages are labeled as no answer.
+            # We want the answer for passage no.3 without regarding the models output for the other passages.
+            if np.sum(idx_text_answers) > 0:
+                group = group.loc[idx_text_answers,:]
+            else:
+                logger.info(f"No textual prediction found in doc: {uid}")
+            if self.top_n_predictions == 1:
+                max_pred = group.loc[group.logit_sum == np.max(group.logit_sum),:]
+                if (max_pred.shape[0] > 1):
+                    max_pred = max_pred.iloc[0, :]
+                    logger.info(f"Multiple predictions have the exact same probability of occuring: \n{max_pred.head()}")
+            else:
+                assert isinstance(self.top_n_predictions, int)
+                sorted_group = group.sort_values(by="logit_sum",ascending=False)
+                filtered_group = sorted_group.drop_duplicates(keep="first", subset="answer_indices")
+                max_pred = filtered_group.iloc[:self.top_n_predictions,:]
+            max_per_sample.append(max_pred.values)
+        return max_per_sample
+
+
+    def _get_best_textanswer_indices(self, logits, n_best_size):
+        """Get the n-best logits from a numpy array without considering the zero index.
+        zero index corresponds to no answer, which we deal with separately"""
+        logits_without_zero = logits[:,1:]
+        idx_without_zero = np.argsort(logits_without_zero,axis=1)[:,-n_best_size:]
+        idx = idx_without_zero + 1
         return idx
+
+class SquadHead(QuestionAnsweringHead):
```

Comment:
Why do we need this extra head? Can't we move this load() method into QuestionAnsweringHead itself?